File: 오완진/프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.Arrays;
import java.util.PriorityQueue;

class Solution {
	public int solution(int[][] jobs) {

		// 1. 수행시간이 짧은 작업부터 처리
		// 2. 단, 앞선 작업들이 끝난 후에 처리되도록

		Arrays.sort(jobs, (o1, o2) -> o1[0] - o2[0]);

		PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);

		int N = jobs.length;
		int time = 0;
		int idx = 0;
		int sum = 0;
		
		while (idx < N || !pq.isEmpty()) {
			while (idx < N && jobs[idx][0] <= time)
				pq.add(jobs[idx++]);
			
			if (!pq.isEmpty()) {
				int[] job = pq.poll();
				time += job[1];
				sum += time - job[0];
			} else
				time = jobs[idx][0];
		}
		
		return sum / N;
	}
}

File: 송인재/프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.kt
import java.util.*;

class Solution {
    class Job(val start: Int, val process: Int)
    
    fun solution(jobs: Array<IntArray>): Int {
        var estimatedEnd = 0
        var total = 0
        val pq = PriorityQueue<Job>( compareBy { it.process } )
        
        jobs.sortWith( compareBy<IntArray> { it[0] }.thenBy { it[1] } )
        
        for (job in jobs) {
            while (!pq.isEmpty() && estimatedEnd < job[0]) {
                val working = pq.poll()
                estimatedEnd = Math.max(estimatedEnd, working.start) + working.process
                total += estimatedEnd - working.start
            }
            
            pq.add(Job(job[0], job[1]))
        }
        
        while (!pq.isEmpty()) {
            val working = pq.poll()
            estimatedEnd = Math.max(estimatedEnd, working.start) + working.process;
            total += estimatedEnd - working.start;
        }
        
        return total / jobs.size
    }
}